{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c7a0b6f9-b392-49fe-afb6-7b227bf5a73d",
   "metadata": {},
   "source": [
    "# Predicting the effect of combinatorial drug treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc061b8c-aaab-413f-8f20-a0f07c812dde",
   "metadata": {},
   "source": [
    "In this tutorial, we predict the effect of combinatorial drug treatments on the A549 cell line. We will also look at the effect of the encoder-decoder module. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52cbffa5-7171-4c5e-8d21-c188af805502",
   "metadata": {},
   "source": [
    "## Preliminaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07690e5c-523c-4a81-93c8-d57f89cabc07",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/icb/dominik.klein/mambaforge/envs/cellflow/lib/python3.12/site-packages/optuna/study/_optimize.py:29: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from optuna import progress_bar as pbar_module\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "from pandas.errors import SettingWithCopyWarning\n",
    "import matplotlib\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "warnings.simplefilter(\"ignore\", UserWarning)\n",
    "warnings.simplefilter(\"ignore\", FutureWarning)\n",
    "warnings.simplefilter(\"ignore\", SettingWithCopyWarning)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import jax\n",
    "from tqdm import tqdm\n",
    "import functools\n",
    "import matplotlib.pyplot as plt\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import rapids_singlecell as rsc\n",
    "import flax.linen as nn\n",
    "import optax\n",
    "import pertpy\n",
    "import cellflow\n",
    "from cellflow.model import CellFlow\n",
    "import cellflow.preprocessing as cfpp\n",
    "from cellflow.utils import match_linear\n",
    "from cellflow.plotting import plot_condition_embedding\n",
    "from cellflow.preprocessing import transfer_labels, compute_wknn, centered_pca, project_pca, reconstruct_pca, annotate_compounds, get_molecular_fingerprints\n",
    "from cellflow.metrics import compute_r_squared, compute_e_distance\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9fb9eba-86e1-45fb-96bb-81f10e59cfad",
   "metadata": {},
   "source": [
    "We download the data from [pertpy](https://pertpy.readthedocs.io/en/stable/):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb05381d-1f6a-4907-9996-95ee2236495c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>n.umi</th>\n",
       "      <th>RT_well</th>\n",
       "      <th>Drug1</th>\n",
       "      <th>Drug2</th>\n",
       "      <th>Well</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>leiden</th>\n",
       "      <th>condition</th>\n",
       "      <th>pathway1</th>\n",
       "      <th>pathway2</th>\n",
       "      <th>split</th>\n",
       "      <th>control</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_18</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.533816</td>\n",
       "      <td>1433</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1004</td>\n",
       "      <td>1004</td>\n",
       "      <td>1433.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>3.558967</td>\n",
       "      <td>1</td>\n",
       "      <td>control+Panobinostat</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>HDAC inhibitor</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_25</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.539404</td>\n",
       "      <td>1448</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1062</td>\n",
       "      <td>1062</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>3.660221</td>\n",
       "      <td>1</td>\n",
       "      <td>control+Panobinostat</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>HDAC inhibitor</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_48</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.532326</td>\n",
       "      <td>1429</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1063</td>\n",
       "      <td>1063</td>\n",
       "      <td>1429.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>2.729181</td>\n",
       "      <td>1</td>\n",
       "      <td>control+Panobinostat</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>HDAC inhibitor</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_51</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>1.461754</td>\n",
       "      <td>3924</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>2412</td>\n",
       "      <td>2411</td>\n",
       "      <td>3923.0</td>\n",
       "      <td>215.0</td>\n",
       "      <td>5.480500</td>\n",
       "      <td>1</td>\n",
       "      <td>control+Panobinostat</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>HDAC inhibitor</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_70</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.418708</td>\n",
       "      <td>1124</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>834</td>\n",
       "      <td>832</td>\n",
       "      <td>1122.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>4.278075</td>\n",
       "      <td>1</td>\n",
       "      <td>control+Panobinostat</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>HDAC inhibitor</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   sample  Size_Factor  n.umi RT_well  \\\n",
       "Cell                                                                    \n",
       "A01_A02_RT_BC_10_Lig_BC_18  sciPlex_theis     0.533816   1433   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_25  sciPlex_theis     0.539404   1448   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_48  sciPlex_theis     0.532326   1429   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_51  sciPlex_theis     1.461754   3924   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_70  sciPlex_theis     0.418708   1124   RT_10   \n",
       "\n",
       "                              Drug1         Drug2 Well  n_genes  \\\n",
       "Cell                                                              \n",
       "A01_A02_RT_BC_10_Lig_BC_18  control  Panobinostat  A10     1004   \n",
       "A01_A02_RT_BC_10_Lig_BC_25  control  Panobinostat  A10     1062   \n",
       "A01_A02_RT_BC_10_Lig_BC_48  control  Panobinostat  A10     1063   \n",
       "A01_A02_RT_BC_10_Lig_BC_51  control  Panobinostat  A10     2412   \n",
       "A01_A02_RT_BC_10_Lig_BC_70  control  Panobinostat  A10      834   \n",
       "\n",
       "                            n_genes_by_counts  total_counts  total_counts_mt  \\\n",
       "Cell                                                                           \n",
       "A01_A02_RT_BC_10_Lig_BC_18               1004        1433.0             51.0   \n",
       "A01_A02_RT_BC_10_Lig_BC_25               1062        1448.0             53.0   \n",
       "A01_A02_RT_BC_10_Lig_BC_48               1063        1429.0             39.0   \n",
       "A01_A02_RT_BC_10_Lig_BC_51               2411        3923.0            215.0   \n",
       "A01_A02_RT_BC_10_Lig_BC_70                832        1122.0             48.0   \n",
       "\n",
       "                            pct_counts_mt leiden             condition  \\\n",
       "Cell                                                                     \n",
       "A01_A02_RT_BC_10_Lig_BC_18       3.558967      1  control+Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_25       3.660221      1  control+Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_48       2.729181      1  control+Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_51       5.480500      1  control+Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_70       4.278075      1  control+Panobinostat   \n",
       "\n",
       "                           pathway1        pathway2  split  control cell_type  \n",
       "Cell                                                                           \n",
       "A01_A02_RT_BC_10_Lig_BC_18  Vehicle  HDAC inhibitor  train        0      A549  \n",
       "A01_A02_RT_BC_10_Lig_BC_25  Vehicle  HDAC inhibitor  train        0      A549  \n",
       "A01_A02_RT_BC_10_Lig_BC_48  Vehicle  HDAC inhibitor  train        0      A549  \n",
       "A01_A02_RT_BC_10_Lig_BC_51  Vehicle  HDAC inhibitor  train        0      A549  \n",
       "A01_A02_RT_BC_10_Lig_BC_70  Vehicle  HDAC inhibitor  train        0      A549  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = pertpy.data.combosciplex()\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0557a706-93c6-41b1-9d79-482ae431b88b",
   "metadata": {},
   "source": [
    "Let's have a look at the most relevant columns:\n",
    "- {attr}`obs['Drug1'] <anndata.AnnData.obs>` and {attr}`obs['Drug2'] <anndata.AnnData.obs>` contain the drug treatments. If the treatment consists of a single drug perturbation, {attr}`obs['Drug2'] <anndata.AnnData.obs>` contains `\"control\"`.\n",
    "- {attr}`obs['cell_type'] <anndata.AnnData.obs>` denotes the cell line. As all treatments are applied to A549 cell populations, we don't need to take this into account. Similarly, the dosage is constant across treatments, and hence can be ignored."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a402154-5989-431e-9bde-98720567a76a",
   "metadata": {},
   "source": [
    "Thus, the experimental condition is uniquely defined by the combination of drugs. We adapt the {attr}`obs['condition'] <anndata.AnnData.obs>` slightly to facilitate the preprocessing part:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "60ae0a73-70f0-4a11-b43c-e4b69b43abcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"condition\"] = adata.obs.apply(\n",
    "    lambda x: \"control\" if x[\"condition\"] == \"control+control\" else x[\"condition\"], axis=1\n",
    ")\n",
    "\n",
    "adata.obs[\"is_control\"] = adata.obs.apply(\n",
    "    lambda x: True if x[\"condition\"] == \"control\" else False, axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82a74985-1d8b-43e5-97a6-a3cf220e33e5",
   "metadata": {},
   "source": [
    "We require embeddings for the drugs. While we encourage users to try different ones, we use molecular fingerprints in the following. Therefore, we first annotate the drugs, i.e. we retrieve the SMILES and [PubChem](https://pubchem.ncbi.nlm.nih.gov/) metadata using {func}`~cellflow.preprocessing.annotate_compounds`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e6bef794-8082-4e50-8b55-506222f1db21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>n.umi</th>\n",
       "      <th>RT_well</th>\n",
       "      <th>Drug1</th>\n",
       "      <th>Drug2</th>\n",
       "      <th>Well</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>split</th>\n",
       "      <th>control</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>is_control</th>\n",
       "      <th>Drug1_pubchem_name</th>\n",
       "      <th>Drug1_pubchem_ID</th>\n",
       "      <th>Drug1_smiles</th>\n",
       "      <th>Drug2_pubchem_name</th>\n",
       "      <th>Drug2_pubchem_ID</th>\n",
       "      <th>Drug2_smiles</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_18</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.533816</td>\n",
       "      <td>1433</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1004</td>\n",
       "      <td>1004</td>\n",
       "      <td>1433.0</td>\n",
       "      <td>...</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_25</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.539404</td>\n",
       "      <td>1448</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1062</td>\n",
       "      <td>1062</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>...</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_48</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.532326</td>\n",
       "      <td>1429</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>1063</td>\n",
       "      <td>1063</td>\n",
       "      <td>1429.0</td>\n",
       "      <td>...</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_51</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>1.461754</td>\n",
       "      <td>3924</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>2412</td>\n",
       "      <td>2411</td>\n",
       "      <td>3923.0</td>\n",
       "      <td>...</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_10_Lig_BC_70</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.418708</td>\n",
       "      <td>1124</td>\n",
       "      <td>RT_10</td>\n",
       "      <td>control</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>A10</td>\n",
       "      <td>834</td>\n",
       "      <td>832</td>\n",
       "      <td>1122.0</td>\n",
       "      <td>...</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   sample  Size_Factor  n.umi RT_well  \\\n",
       "Cell                                                                    \n",
       "A01_A02_RT_BC_10_Lig_BC_18  sciPlex_theis     0.533816   1433   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_25  sciPlex_theis     0.539404   1448   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_48  sciPlex_theis     0.532326   1429   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_51  sciPlex_theis     1.461754   3924   RT_10   \n",
       "A01_A02_RT_BC_10_Lig_BC_70  sciPlex_theis     0.418708   1124   RT_10   \n",
       "\n",
       "                              Drug1         Drug2 Well  n_genes  \\\n",
       "Cell                                                              \n",
       "A01_A02_RT_BC_10_Lig_BC_18  control  Panobinostat  A10     1004   \n",
       "A01_A02_RT_BC_10_Lig_BC_25  control  Panobinostat  A10     1062   \n",
       "A01_A02_RT_BC_10_Lig_BC_48  control  Panobinostat  A10     1063   \n",
       "A01_A02_RT_BC_10_Lig_BC_51  control  Panobinostat  A10     2412   \n",
       "A01_A02_RT_BC_10_Lig_BC_70  control  Panobinostat  A10      834   \n",
       "\n",
       "                            n_genes_by_counts  total_counts  ...  split  \\\n",
       "Cell                                                         ...          \n",
       "A01_A02_RT_BC_10_Lig_BC_18               1004        1433.0  ...  train   \n",
       "A01_A02_RT_BC_10_Lig_BC_25               1062        1448.0  ...  train   \n",
       "A01_A02_RT_BC_10_Lig_BC_48               1063        1429.0  ...  train   \n",
       "A01_A02_RT_BC_10_Lig_BC_51               2411        3923.0  ...  train   \n",
       "A01_A02_RT_BC_10_Lig_BC_70                832        1122.0  ...  train   \n",
       "\n",
       "                            control cell_type is_control Drug1_pubchem_name  \\\n",
       "Cell                                                                          \n",
       "A01_A02_RT_BC_10_Lig_BC_18        0      A549      False            control   \n",
       "A01_A02_RT_BC_10_Lig_BC_25        0      A549      False            control   \n",
       "A01_A02_RT_BC_10_Lig_BC_48        0      A549      False            control   \n",
       "A01_A02_RT_BC_10_Lig_BC_51        0      A549      False            control   \n",
       "A01_A02_RT_BC_10_Lig_BC_70        0      A549      False            control   \n",
       "\n",
       "                           Drug1_pubchem_ID Drug1_smiles  Drug2_pubchem_name  \\\n",
       "Cell                                                                           \n",
       "A01_A02_RT_BC_10_Lig_BC_18             <NA>          NaN        Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_25             <NA>          NaN        Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_48             <NA>          NaN        Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_51             <NA>          NaN        Panobinostat   \n",
       "A01_A02_RT_BC_10_Lig_BC_70             <NA>          NaN        Panobinostat   \n",
       "\n",
       "                           Drug2_pubchem_ID  \\\n",
       "Cell                                          \n",
       "A01_A02_RT_BC_10_Lig_BC_18          6918837   \n",
       "A01_A02_RT_BC_10_Lig_BC_25          6918837   \n",
       "A01_A02_RT_BC_10_Lig_BC_48          6918837   \n",
       "A01_A02_RT_BC_10_Lig_BC_51          6918837   \n",
       "A01_A02_RT_BC_10_Lig_BC_70          6918837   \n",
       "\n",
       "                                                               Drug2_smiles  \n",
       "Cell                                                                         \n",
       "A01_A02_RT_BC_10_Lig_BC_18  CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO  \n",
       "A01_A02_RT_BC_10_Lig_BC_25  CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO  \n",
       "A01_A02_RT_BC_10_Lig_BC_48  CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO  \n",
       "A01_A02_RT_BC_10_Lig_BC_51  CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO  \n",
       "A01_A02_RT_BC_10_Lig_BC_70  CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotate_compounds(adata, compound_keys=[\"Drug1\", \"Drug2\"])\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7438db45-50af-4c97-8be9-c546a0982f70",
   "metadata": {},
   "source": [
    "Among others, this gave us the SMILES strings, such that we can now get the molecular fingerprints for the SMILES strings using {func}`~cellflow.preprocessing.get_molecular_fingerprints`. We have {attr}`uns['fingerprints'] <anndata.AnnData.uns>` added, and see that all drugs have been assigned a fingerprint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aa4cc50e-e738-464c-acc0-6f5ed8a56bbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Cediranib', 'Givinostat', 'Panobinostat', 'Dacinostat', 'Alvespimycin', 'SRT2104', 'SRT3025', 'PCI-34051', 'SRT1720', 'Sorafenib', 'Danusertib', 'Dasatinib', 'Tanespimycin', 'Carmofur', 'Crizotinib', 'Pirarubicin', 'Curcumin'])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_molecular_fingerprints(adata, compound_keys=[\"Drug1\", \"Drug2\"])\n",
    "adata.uns[\"fingerprints\"].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abe53039-afc8-42fc-a255-40c94d8e79b4",
   "metadata": {},
   "source": [
    "We now add a zero token which is going to be ignored during training for \"filling\" the second drug in case of single drug perturbations. Note that this zero token will be specified later in {meth}`~cellflow.model.CellFlow.prepare_data`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fab30ef8-381e-4ee4-a260-96bc596b9916",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[\"fingerprints\"][\"control\"] = np.zeros_like(next(iter(adata.uns[\"fingerprints\"].values())))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d17fa008-a9b4-4bf1-8b7c-e01b27456f00",
   "metadata": {},
   "source": [
    "We now perform standard preprocessing (note that in the manuscript the preprocessing was slightly different to make it comparable to count-based models):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "41e88b65-4cf7-47dd-845b-5f8eee320c42",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"counts\"].copy()\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, inplace=True, n_top_genes=2000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358b9e08-9497-456f-9e06-fc422c5885c9",
   "metadata": {},
   "source": [
    "We now define a train-test split as in the CellFlow manuscript:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f756deca-b6bf-4fbb-9177-6d857928f1bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_conditions = ['Panobinostat+Crizotinib', 'Panobinostat+Curcumin', 'Panobinostat+SRT1720', 'Panobinostat+Sorafenib', 'SRT2104+Alvespimycin', 'control+Alvespimycin', 'control+Dacinostat']\n",
    "adata.obs[\"mode\"] = adata.obs.apply(lambda x: \"test\" if x[\"condition\"] in test_conditions else \"train\", axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "492f0d90-039c-435f-9dcc-68a8b065365a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2025-05-23 13:10:21.428] [CUML] [debug] n_neighbors=15\n",
      "[2025-05-23 13:10:21.428] [CUML] [debug] Calling knn graph run\n",
      "[2025-05-23 13:10:21.428] [CUML] [debug] Done. Calling fuzzy simplicial set\n",
      "[2025-05-23 13:10:22.774] [CUML] [debug] Done. Calling remove zeros\n",
      "WARNING: .obsp[\"connectivities\"] have not been computed using umap\n"
     ]
    }
   ],
   "source": [
    "rsc.pp.pca(adata)\n",
    "rsc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=[\"mode\", \"condition\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c345e280-a420-486f-9ae0-b99603feb573",
   "metadata": {},
   "source": [
    "We now compute the PCA embedding based on the training data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "77234ec3-cf38-4e4d-ab34-e7fad162f3b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_train = adata[adata.obs[\"mode\"] == \"train\"]\n",
    "adata_test = adata[(adata.obs[\"mode\"] == \"test\") | (adata.obs[\"condition\"]==\"control\")]\n",
    "centered_pca(adata_train, method=\"rapids\", keep_centered_data=False, n_comps=100)\n",
    "project_pca(adata_test, ref_adata=adata_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fdfd08e-3793-44de-8a4f-456fa2d9bdcf",
   "metadata": {},
   "source": [
    "## Setting up the CellFlow model\n",
    "\n",
    "We are now ready to setup the {class}`~cellflow.model.CellFlow` model.\n",
    "\n",
    "Therefore, we first choose the flow matching solver. We select the default solver `\"otfm\"`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "439abf26-eda6-4f0f-991d-dc850f5bdefa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cf = CellFlow(adata_train, solver=\"otfm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1500afe-18b6-4d18-aa6a-91451548cca4",
   "metadata": {},
   "source": [
    "## Preparing {class}`~cellflow.model.CellFlow`'s data handling with {meth}`~cellflow.model.CellFlow.prepare_data`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc10beb6-4401-4b68-8447-69591c393627",
   "metadata": {},
   "source": [
    "- We use {attr}`obsm['X_pca'] <anndata.AnnData.obsm>` as cellular representation.\n",
    "- {attr}`obs['is_control'] <anndata.AnnData.obs>` indicates whether a cell is a control cell.\n",
    "- `perturbation_covariates` indicates the external intervention, which is the genetic perturbation. As we also have combinations thereof, we pass the column names `\"gene_target_1\"` and `\"gene_target_2\"`. We call this intervention `\"genetic_perturbation\"`. We use ESM2 embeddings for representing the cytokines, which we have precomputed already for the purpose of this notebook, saved in {attr}`uns['gene_embeddings'] <anndata.AnnData.uns>`. Thus, we pass the information that `\"gene_embeddings\"` stores embeddings of the {attr}`obs['cytokine'] <anndata.AnnData.obs>` treatments via `perturbation_covariate_reps`.\n",
    "- The sample covariate describes the cellular context independent of the perturbation. We pass the time point via {attr}`obs['logtime'] <anndata.AnnData.obs>`. \n",
    "- The `split_covariates` defines classes within which me map. As we want to learn maps from control to perturbed states *within* the same time point, we set `\"timepoint\"` to be the split covariate. See {doc}`201_zebrafish_continuous` for a different use of `split_covariates`.\n",
    "-  `max_combination_length` is the maximum number of gene knockouts seen during training or inference. We set it to `2`. Note that this value is also automatically inferred from the maximum number of elements seen in `perturbation_covariates`, but can optionally set to be higher for inference.\n",
    "-  The `null_value` is the value which fills the absence of a perturbation, i.e. the value which fills all positions in {attr}`uns['gene_embeddings']['control'] <anndata.AnnData.uns>`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "84d1797f-4671-4a41-aeb3-978d91a8b127",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25/25 [00:01<00:00, 14.49it/s]\n"
     ]
    }
   ],
   "source": [
    "cf.prepare_data(\n",
    "    sample_rep = \"X_pca\",\n",
    "    control_key = \"is_control\",\n",
    "    perturbation_covariates = {\"drug_perturbation\": (\"Drug1\" , \"Drug2\")},\n",
    "    perturbation_covariate_reps = {\"drug_perturbation\": \"fingerprints\"},\n",
    "    max_combination_length = 2,\n",
    "    null_value = 0.0,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1fc1515-30d3-4ee9-92bb-0c8299f94d21",
   "metadata": {},
   "source": [
    "We now prepare the data validation data using {meth}`~cellflow.model.CellFlow.prepare_validation_data`. \n",
    "\n",
    "As for some conditions, and in particular for control cells, we have a large number of measurements, we subsample for inference to be faster. However, due to the heterogeneity of the cellular distribution, covering hundreds of cell types, we should not subsample by too much."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5a37f4b0-c165-4c45-b1f0-3b6c7755b26f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25/25 [00:00<00:00, 278.47it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 538.90it/s]\n"
     ]
    }
   ],
   "source": [
    "cf.prepare_validation_data(\n",
    "    adata_train,\n",
    "    name=\"train\",\n",
    "    n_conditions_on_log_iteration=None,\n",
    "    n_conditions_on_train_end=None,\n",
    ")\n",
    "\n",
    "cf.prepare_validation_data(\n",
    "    adata_test,\n",
    "    name=\"test\",\n",
    "    n_conditions_on_log_iteration=None,\n",
    "    n_conditions_on_train_end=None,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806d7551-1a1a-4080-abfc-d8839724d7a2",
   "metadata": {},
   "source": [
    "## Preparing {class}`~cellflow.model.CellFlow`'s model architecture with {meth}`~cellflow.model.CellFlow.prepare_model`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f0ed52-4cd9-45c6-af62-3b230a49903c",
   "metadata": {},
   "source": [
    "We are now ready to specify the architecture of {class}`~cellflow.model.CellFlow`.\n",
    "\n",
    "We only consider the most relevant parameters, for a detailed description, please have a look at the documentation of {meth}`~cellflow.model.CellFlow.prepare_model` or {doc}`100_pbmc`.\n",
    "\n",
    "- We use `condition_mode=\"deterministic\"` to learn point estimates of condition embeddings, and thus have a fully deterministic mapping. We set `regularization=0.0`, thus don't regularize the learnt latent space. \n",
    "- `pooling` defines how we aggregate combinations of conditions in a permutation-invariant manner,  which we choose to do learning a class token indicated by `\"attention_token\"`.\n",
    "- With `layers_before_pool` we define how to embed the ESM2 embeddings and the log-value of the time. We process them such that have the same scale of dimensionality.\n",
    "- `condition_embedding_dim` is the dimension of the latent space of the condition encoder.\n",
    "- `cond_output_dropout` is the dropout applied to the condition embedding, we recommend to set it relatively high, especially if the `condition_embedding_dim` is large.\n",
    "- `pool_sample_covariates` defines whether the concatenation of the sample covariates should happen before or after pooling, in our case indicating whether it's part of the self-attention or only appended afterwards. \n",
    "- `flow` defines the reference vector field between pairs of samples which the {class}`~cellflow.networks._velocity_field.ConditionalVelocityField` is regressed against. Here, we use `{\"constant_noise\": 0.5}`, i.e. we use a relatively small value as we have a highly heterogeneous cell population. In fact, if we augment a cell with noise, we should be careful not to augment it to the extent that it is e.g. in a completely different organ of the zebrafish.\n",
    "- `match_fn` defines how to sample pairs between the control and the perturbed cells. As we have a strongly heterogeneous population, we choose a higher batch size of 2048. We don't expect large outliers, and are not interested in the trajectory of a single cell, hence we choose `tau_a=tau_b=1.0`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c97455f-efd0-4b30-b6e5-70c82cf472fb",
   "metadata": {},
   "source": [
    "We start with defining the `layers_before_pool` and `layers_after_pool`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "872cd261-882c-42d2-8f4f-f9eb81c7f85d",
   "metadata": {},
   "outputs": [],
   "source": [
    "layers_before_pool = {\n",
    "    \"drug_perturbation\": {\"layer_type\": \"mlp\", \"dims\": [256, 256], \"dropout_rate\": 0.0},\n",
    "}\n",
    "\n",
    "layers_after_pool = {\n",
    "    \"layer_type\": \"mlp\", \"dims\": [256, 256], \"dropout_rate\": 0.0,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43c4f20f-3b20-4196-b593-691df8a8159a",
   "metadata": {},
   "source": [
    "We now also explicitly define the `match_fn`. As the data set contains a few outliers, we set `tau_b=0.95`, thus allowing for some unbalancedness in the perturbed populations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "95a8d037-1669-419a-80ba-d11d2b5768cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_fn = functools.partial(match_linear, epsilon=1.0, tau_a=1.0, tau_b=1.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecb0a272-6bab-4fc0-b316-b552be6641e9",
   "metadata": {},
   "source": [
    "Now we are ready to prepare the model. Note that one parameter stands out in this case, namely `flow={\"constant_noise\": 1.5}` has a relatively high noise value. While the magnitude of noise depends on the scale of the cell embeddings (and thus is relatively high in PCA space, e.g. compared to VAE space), it also depends on the type of data. In cell lines, we can normally choose a relatively high noise value on the probability path, while on heterogeneous tissue, we have to select a lower magnitude"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "55e3d8ce-ce59-479a-9276-9c3432ddb2c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cf.prepare_model(\n",
    "    condition_mode=\"deterministic\",\n",
    "    regularization=0.0,\n",
    "    pooling=\"mean\",\n",
    "    layers_before_pool=layers_before_pool,\n",
    "    layers_after_pool=layers_after_pool,\n",
    "    condition_embedding_dim=64,\n",
    "    cond_output_dropout=0.9,\n",
    "    hidden_dims=[2048, 2048, 2048],\n",
    "    conditioning=\"concatenation\",\n",
    "    decoder_dims=[4096, 4096, 4096],\n",
    "    probability_path={\"constant_noise\": 1.5},\n",
    "    match_fn=match_fn,\n",
    "    linear_projection_before_concatenation=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c60ee3e-a325-4f80-95ff-f6e506552a20",
   "metadata": {},
   "source": [
    "## Computing and logging metrics during training \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e5508d-b8f6-4d8e-978b-20c557557bf7",
   "metadata": {},
   "source": [
    "We compute the energy distance and MMD during training, and the R squared in gene expression space. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4134dbb8-dea3-41ff-ba5b-cba755aeb40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics_callback = cellflow.training.Metrics(metrics=[\"mmd\", \"e_distance\"])\n",
    "decoded_metrics_callback = cellflow.training.PCADecodedMetrics(ref_adata=adata_train, metrics=[\"r_squared\"])\n",
    "callbacks = [metrics_callback, decoded_metrics_callback]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "503fb8d8-909d-419f-85cb-18abf469c0c0",
   "metadata": {},
   "source": [
    "## Training CellFlow\n",
    "\n",
    "Finally, we are ready to train our model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6086daf9-c4ec-4327-a28f-55b86aaa287e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 300000/300000 [1:18:08<00:00, 63.99it/s, loss=0.631]    \n"
     ]
    }
   ],
   "source": [
    "cf.train(\n",
    "        num_iterations=300_000,\n",
    "        batch_size=1024,\n",
    "        callbacks=callbacks,\n",
    "        valid_freq=60_000,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44171594-47e4-458d-8e29-34c3d5e2979f",
   "metadata": {},
   "source": [
    "We can now investigate some training statistics, stored by the {class}`~cellflow.training.CellFlowTrainer`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "11336bcc-1515-4f4f-b9db-a9d3f39d0acc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['loss', 'train_mmd_mean', 'test_mmd_mean', 'train_e_distance_mean', 'test_e_distance_mean', 'pca_decoded_test_r_squared_mean', 'pca_decoded_train_r_squared_mean'])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cf.trainer.training_logs.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "81d0c4d1-f947-4cc4-8fc8-7a1257d24186",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_distances_train = cf.trainer.training_logs[\"train_e_distance_mean\"]\n",
    "e_distances_test = cf.trainer.training_logs[\"test_e_distance_mean\"]\n",
    "pca_decoded_test_r_squared_mean = cf.trainer.training_logs[\"pca_decoded_test_r_squared_mean\"]\n",
    "\n",
    "iterations_train = np.arange(len(e_distances_train))\n",
    "iterations_test = np.arange(len(e_distances_test))\n",
    "iterations_r2 = np.arange(len(pca_decoded_test_r_squared_mean))\n",
    "\n",
    "fig, axes = plt.subplots(1, 3, figsize=(18, 6))\n",
    "\n",
    "axes[0].plot(iterations_train, e_distances_train, linestyle='-', color='blue', label='Energy distance (train)')\n",
    "axes[0].set_xlabel('Validation Iteration')\n",
    "axes[0].set_ylabel('Energy distance')\n",
    "axes[0].legend()\n",
    "axes[0].grid(True)\n",
    "\n",
    "axes[1].plot(iterations_test, e_distances_test, linestyle='-', color='red', label='Energy distance (test)')\n",
    "axes[1].set_xlabel('Validation iteration')\n",
    "axes[1].set_ylabel('Energy distance')\n",
    "axes[1].tick_params(axis='y')\n",
    "axes[1].legend(loc='upper left')\n",
    "axes[1].grid(True)\n",
    "\n",
    "axes[2].plot(iterations_test, pca_decoded_test_r_squared_mean, linestyle='-', color='red', label='Decoded test R squared')\n",
    "axes[2].set_xlabel('Validation iteration')\n",
    "axes[2].set_ylabel('R squared')\n",
    "axes[2].tick_params(axis='y')\n",
    "axes[2].legend(loc='upper left')\n",
    "axes[2].grid(True)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edeb6540-e54b-4746-afcf-49a8478c824f",
   "metadata": {},
   "source": [
    "## Investigating the learnt latent space"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1e33895-4d76-4113-97b1-8f46a3de9037",
   "metadata": {},
   "source": [
    "We can visualize the learnt latent space for using {meth}`~CellFlow.get_condition_embedding`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7c83aa9b-7da8-4988-bdce-741f909edeb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [00:00<00:00, 1121.42it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 1025.25it/s]\n"
     ]
    }
   ],
   "source": [
    "covariate_data_train = adata_train[~adata_train.obs[\"is_control\"].to_numpy()].obs.drop_duplicates(subset=[\"condition\"])\n",
    "covariate_data_test = adata_test[~adata_test.obs[\"is_control\"].to_numpy()].obs.drop_duplicates(subset=[\"condition\"])\n",
    "\n",
    "df_embedding_train = cf.get_condition_embedding(covariate_data=covariate_data_train, condition_id_key=\"condition\", rep_dict=adata_train.uns)[0]\n",
    "df_embedding_test = cf.get_condition_embedding(covariate_data=covariate_data_test, condition_id_key=\"condition\", rep_dict=adata_train.uns)[0]\n",
    "\n",
    "df_embedding_train[\"seen_during_training\"] = True\n",
    "df_embedding_test[\"seen_during_training\"] = False\n",
    "df_condition_embedding = pd.concat((df_embedding_train, df_embedding_test))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "67b5b57e-e994-4177-a59e-8e3bb46d577e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAdQAAAGDCAYAAACSrOEjAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAALDpJREFUeJzt3Xl4VOXB/vF7skF2iAYSCJsoEColkEBYtSKbQmS1tShaaCMUFaiKooBAfBGVom/0rQIWDKUVURBRcSUYFPTHEqgIhsgui2ERgSxknfn9ERmN2SbJmTmT4fu5Lq5mznkm554UuXOes1lsNptNAACgTrzMDgAAgCegUAEAMACFCgCAAShUAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQpUY+zYsbr++uvVpUsXde/eXWPHjtWePXskSfn5+Xr++ec1cOBAxcTE6IYbbtDkyZPt6y97++231b59e7322mtmfAQALkChAg54+OGHtWvXLn3++eeKjo7WpEmTVFRUpPHjx2vbtm16/vnntX37dn344YcaMGCAPvnkkzLvX716tRo1aqTVq1eb9AkAOJuP2QGA+qRBgwYaPXq0li9frvfee08HDx7UJ598opCQEEmSr6+vEhISyrznyJEj2r59u/7xj3/o/vvv1759+9ShQwcz4gNwIvZQgRq4dOmS3nzzTTVv3lybN2/WDTfcYC/TyqxZs0YdO3ZU//79FRcXx14q4KEoVMABzz33nOLi4tS/f38dOnRIL730ks6dO6emTZtW+b6SkhKtXbtWw4cPlyQNGzZM7777rgoLC12QGoArUaiAAx588EHt2LFDW7Zs0dKlS9WhQwc1btxYp06dqvJ9mzZt0vnz5+3TwLfccovy8/PLHWMFUP9RqEAt9enTR59//rmys7MrHbN69WpZrVYlJCSod+/eGjRokIqLi5n2BTwQJyUBtZSQkKDVq1dr4sSJmjFjhtq1a6fi4mJt3LhR3377re666y5t2rRJzzzzjHr06GF/X0ZGhu69914dP35cUVFRJn4CAEaiUIFa8vX11dKlS7Vo0SJNmTJFp0+fVmhoqGJiYjRhwgStXbtWkZGRGjJkiLy8fp4MCg8PV8eOHbVmzRpNmTLFxE8AwEgWm81mMzsEAAD1HcdQAQAwAIUKAIABKFQAAAxAoQIAYAAKFQAAA1CoAAAYwOML1WazKScnR1wdBABwJo8v1NzcXMXGxio3N9fsKAAAD+bxhQoAgCtQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAEoVAAADMADxuGeii5JPx6VfPyksGvMTgMA1aJQ4V7yL0ifzpf++5pUcKF0WXi01HuyFDPG3GwAUAUKFe6jIFtKGSpl7S67/EyG9PZfS/dYb3rMnGwAUA2OocJ9bHmhfJn+0qZnpLP7XZcHAGqAQoV7sFqlnf+qZpBNSk9xRRoAqDEKFe4h/7yUk1X9uDOZTo8CALVBocI9+AZIFu/qx/kFOj8LANQChQr34NtQaje4+nG/GeH8LABQCxQq3EefqZJXFSeeh0dLHYa6LA4A1ASFCvfRors0epnkF1R+XUQn6a41kjdXegFwT/zrBPfScZh0zU3S7lWll9D4/DQV3LafZLGYnQ4AKkWhwv00DJG6J5qdAgBqhClfAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAG4sQPcW0FO6V2Tvv1IKimQIjtLceOlxq3NTgYAZVCocF9ZX0v/Hl32OamH0qQvXpQGPyPF32taNAD4NaZ84Z4Kc8uX6WU2q/TBI9KBVNfnAoBKUKhwT7tXVVymdjbpy/9zWRwAqA6FCvf07cfVjzn4qVRc4PwsAOAAChXuqcSRorRJJUVOjwIAjqBQ4Z4iY6ofc9V1UoMKHkYOACagUOGe4sZJFu+qx3T7s2uyAIADKFS4p0YtpVsXSLJUvP66gVI3HkIOwH1wHSrcV9e7pUs/SnvWSGf26UBJhF71/YM2lHRV4cGG+s2r6bqrR0sNvj7S7KQAQKHCTaWnSBvnSbmnJUkbSrpqUtEUFRb6/jSgSJsPnNXmA2d1e2yUnh39W1kslezNAoALMOUL97N9qfTuFHuZnrMF64Gi+1Uo3wqHv5l+XKu2H3NlQgAoh0KFeynKlzY+WWbRGyU36pIaVvm2lC+OODEUAFSPQoV7yXy/9LjpL2yzdqj2bfuyspWdzzWpAMxDocK9ZJe/3aCXbA69lWOoAMxUrwo1Pz9fAwYMUFxcnNlR4CzBTcst6uW1t9q3dW7RSEENOMcOgHnqVaEmJyerWbNmZseAM7W/VWrYqMyi0d6bFKLcKt82vndr52UCAAfUm0Lds2ePNm/erMRELub3aL7+0k0zyiwKsVzSEr+FClJehW9J7NtGw2KauyIdAFSqXsyRFRcXa9asWXriiSdktVrNjgNni79XsliktPlS3g+SpB5e+/RJ2EK9Fjldn/wQpsJiqzo2C9HYHq0Uf81VJgcGgHpSqEuXLlV0dLS6deumrVu3mh0HrtA9sfROSfs/Kb0eNSRKkW376SFvHz1kdjYAqIDbF+rRo0f1+uuva+3atWZHgav5NJCih5qdAgAc4vaFmp6errNnz2rQoEGSSqd/c3NzFR8fryVLlqhz584mJwQAoB4U6i233KJevXrZX+/atUszZ87UunXrFBYWZmIyAAB+5vaF6u/vL39/f/vrsLAwWSwWRUREmJgKAICy6s1lM5fFx8drx44dZscAAKCMeleoAAC4IwoVAAADUKgAABiAQgUAwAAUKgAABqBQAQAwAIUKAIABKFQAAAxAoQIAYAAKFQAAA1CoAAAYgEIFAMAAFCoAAAagUAEAMACFCgCAAShUAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAEoVAAADEChAgBgAAoVAAADUKgAABiAQgUAwAAUKgAABqBQAQAwAIUKAIABKFQAAAxAoQIAYAAKFQAAA1CoAAAYgEIFAMAAFCoAAAagUAEAMACFCgCAAdy+UAsLCzVz5kz169dPXbp00eDBg7V69WqzYwEAUIaP2QGqU1xcrPDwcKWkpKhFixb66quvlJiYqIiICPXp08fseAAASKoHe6gBAQGaMmWKWrZsKYvFopiYGMXHxys9Pd3saAAA2Ll9of5aQUGBdu/erfbt25sdBQAAu3pVqDabTTNmzFCrVq00cOBAs+MAAGDn9sdQL7PZbJozZ44OHz6slJQUeXnVq98FAAAerl4Uqs1m09y5c7V7926lpKQoODjY7EgAAJRRLwo1KSlJO3fu1PLlyxUaGmp2HAAAynH7Qj1x4oRee+01+fn5qV+/fvblCQkJSkpKMjEZAAA/c/tCbd68uTIzM82OAQBAlTizBwAAA1CoAAAYgEIFAMAAFCoAAAagUAEAMACFCgCAAShUAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADCA2z9tpt7KOydlvCNdOi81biW1HyL5+JmdCgDgJBSq0axWKXWutHWRVJz/8/LAcGnw01Kn0eZlAwA4DYVqtE9mSV/+X/nluWekNX+RfAOkDre6PhcAwKk4hmqk7FPS1sVVDLBJG//HZXEAAK5DoRpp71uStajqMaf3St/vdk0eAIDLUKhGyj3r2LicU87NAQBwOQrVSKFRjo37YLqUneXcLAAAl6JQjXT9SMk3sPpx5w5Iq+5yfh4AgMtQqEZqGCrdPMuxsce3S0e/dG4eAIDLUKhGyjsneftKzbo6Nv7bD5ybBwDgMlyHapRNz0qfLyx7M4fqFBc6Lw8AwKUoVCNsfl76dF7N39csxvAoAABzMOVbV4V5pYVaUwFXSb8ZYXweAIApKNS62v+xlH+hZu/x9pNGviL5NHBOJgCAy1GodXXpx5q/Z8QS6dqbjc8CADANx1DrqnGrmo23eEsXT0g2m2SxOCcTgMrZbNKBVOnQp5LNKrWIlzoMlbw965/DkpISFRVVcytUVMvX11fe3t4OjfWsv0FmaPO70kez5Z5xbLytRPp4hvTjYWnIQmcmA/BrPxyUXh8jndn387L/95IU0lz6/QopKta8bAbKycnR8ePHZbPZzI5S71ksFkVFRSkoKKj6sTYP/4nn5OQoNjZW6enpDv1Ayii6JO1eJf13Zen9d0OaSTFjpE63lz3+uW996X+kNZX4qdTcwWtWAdRN/gXp5d7ShWMVr28YKk3cLDVq6dpcBispKdH+/fsVEBCg8PBwWZgJqzWbzaYzZ84oLy9P1113XbV7quyhVib3rPSvYdKpPT8v+/GwdHSLtGOZdNdbkn+j0uUdhkgDn5Q+fkJSDX4/SU+hUAFX+e9rlZepVFq4WxdLg2pxCZwbKSoqks1mU3h4uPz9/c2OU++Fh4fryJEjKioqqrZQOSmpMuvuK1umv3QiXVr/YNllvSZLd/ynZts4d6h22QDU3J63qlx9yean779K1aXCEhcFci72TI1Rk58je6gV+eGg9O1HVY/5Zp108WTpNPBl4R1qtp0GwTXPBqB2CrIrXHzQGqkXi0fofWt3FRb4yS/pYw3pFKkH+l2ra8JreJgIVzT2UCty5HNVO3VrLZaOflF22fnjNdtO9LCajQdQe1e1Lbdoj7WVRhQm6W1rHxXKT5JUWGzV2l0nNOKlL/TNyYuuTol6jEKtiM1au3HfrK3Zdnw5vgHU1YVLRfrn54c0dulW/XHJ/9PTH+zTsXN55QfG/qncoqlF9+miKn7k4oVLRZr+1m6D09YfRSVWpWac0spt3yk145SKShz8d9HF3nrrLcXFxZkdQ5JBU75Wq1XvvPOOhg8fbsS3M1/LXtWPsXiVXr/2S5VMKVXqwnc1Gw+gjPSjP+rPy7frfN7P11t+eegHvfL5Ic0bfr3u6P6LM3av7S9dP0ras0b5Nl+NL5ymA7aoKr//7uMX9PXxC+oUFeqsj+CW3txxTAs+ytTp7AL7sibBDTRtUHvdHtfCKducPn261q4tv1Py8ccfq1WrGl7vbxJD9lCLi4v12GOPGfGt3EOTDlKbG6oe0+6W8jd1CAyv2XYaXln/kQJG+jG3UONTypbpZSVWmx5f+7W2HT7380KLpfSWn81j9VDRRH1hu96h7WR8f2VN+76545imrd5dpkwl6XR2gaat3q03d1RxpnQd9e3bV5s3by7zJyqq6l963InDe6ipqamVrisuLjYkjFsZvkhaPrTiM3HDO0gJyeWXd7pd2vqyY9/fp0Hp3VkA1MobO47pwqXK7wRktUlLNx9S9zZhpQv2vS9teloHTpzRemtPh7fT0M+xu+R4gqISqxZ8lFnlmAUfZWpEl+by8Tb+iKGfn5/Cw8vumLz66qt66623dOzYMYWGhuqmm27StGnTFBhY8VT9vn37NG/ePO3Zs0cWi0WtW7fW3Llz1alTJ0nSjh079Nxzz2nPnj1q3LixBgwYoAcffFABAQF1zu9wod53332KjIw05VTsoqIizZ8/X++++64sFosSEhL02GOPycfHiScphzYvvfHCzuWl16/lnJaCI6Uud0pd7674DN2oWOnagdKBj6v//t3vlQLCjM8NXCE+zTxd/Zh9P93B7L+vSW9PkmTTeqvjT3ny8bLoxutqOPNUj3327Zlye6a/djq7QJ/tP6N+HZq6JJPFYtGMGTMUFRWlY8eOae7cuVqwYIHmzJlT4fiHH35Y0dHRmjNnjry9vZWRkSFfX19J0nfffafExERNmTJFTz31lM6dO6cnn3xSTz75pObPn1/nrA43UrNmzbRgwQLFxpa/NVdBQYFiYmLqHKYyL7/8stLT07V+/XpJUmJiohYtWqT777/faduUVHrjht5TSv846vcp0qq7pYMbKh/T9R6p/9y6pgOuaEUl1d9EpdhqlS0/W5YPHtXlM/ezbY7vifh6eymo4ZVzdWF1ZWofd9GxcTWVlpamLl262F/37dtXL7zwgv11VFSUpk6dqtmzZ1daqCdPntSf//xntW1belZ369at7esWL16shIQE/elPf7KvmzFjhsaOHas5c+aoQYO6PQHM4b8p119/vfbu3VthoVosFkVGRtYpSFXWrFmjxx57TE2aNJEkTZw4Uc8++6zzC7U2/AKlsWuk4zuljUnSiR1SUb7UMKT0JKYBSdLV15mdEqj3OjUPVfrRqp/2dH3zUFn2viUV/HwctI3le4e3camoROfzCnVV0JXxqMUmwY59ziYhzvl5xMfHlylKf39/ffHFF1q8eLEOHTqknJwclZSUqKCgQJcuXarwTlDjxo3TzJkztW7dOvXq1UuDBw9Wy5alJ6ft27dPmZmZevfdd+3jbTabrFarjh8/bi/h2nK4UBcurPxG7n5+ftq4cWOdglTmwoULysrKUnR0tH1ZdHS0Tp48qezsbAUHu+nNEaK6Sne/bXYKwGPd1aOVln95RFXdjfyuHq2kc2UPwdzm/aWeKr5Tuar+sjWLRWrge+UcQ72hXbiaBDeock+1SXAD3eCkaXB/f/8yZ/QeP35cEyZM0B//+Ef97W9/U2hoqNLT0zVjxgwVFRVVWKgPPPCAhg4dqk2bNumzzz7TCy+8oOeff14DBgxQXl6e7rjjDo0dO7bc+4zYKXT4qLKvr699HtqV8vJKryf7ZXGGhIRIknJzc12eB4B7uLZJkGYO6Vjp+oTOzXR7bFS58x2CLZf0hM8KSdVfV9nn2qsV1ODKmfL19fbStEHtqxzzyOAOTjkhqSJ79+6VzWbT9OnTFRMTozZt2uj06eqPnbdp00Z/+tOftGzZMg0cOFBr1qyRJHXs2FEHDhxQq1atyv3x8/Orc163v7HD5TOvcnJy7Muys0uv96zsLC8AV4Y/92mjf43vrhvbhcvrp/MlO0aGaP7ITkr+Q0zpSZS/GSmp7MmUf/BJ0z98XpCfCiv93l4WmybeWLcpwPro9rgWWjD6t+Wmf5sEN9Dfb++s0bGuu4ylVatWKioq0ooVK3Ts2DG9/fbbev311ysdn5+fr6SkJG3dulUnTpxQenq6vv76a/tUbmJionbt2qWkpCRlZGToyJEj2rBhg5KSkgzJ6/a/eoWGhioiIkIZGRn2efCMjAxFRka673QvAJe5oV24bmgXrhKrTSVWm/x8frWfcFVbqdNo6es3yywe4rNNnb0PaUzB4/pOEWXW+VmsmjcqRr2vvdrZ8d3S7XEtNKJLc322/4xOXyxQk5DSaV5X7Zle1qFDBz322GN65ZVX9NxzzykuLk4PPvigHn300QrHe3l56fz583r00Ud19uxZNW7cWAMHDtTkyZPt32/FihX63//9X40ZU/rIzRYtWujWW281JG+9eB5qcnKy0tLStGTJEknSvffeq5tvvtmhk5Lq9DxUAJ6hKF96+6/S3vJPnCmxWbTBGqsN1q4qaNhE0de21e8ThtTbE5Hy8/N1+PBhtWnTRg0bNjQ7Tr1Xk5+n2++hStKkSZN0/vx5+28Rt912myZOnGhyKgD1hm9D6fZXpd9Nl/asKX32aeM20rX95e3rr0GB4Rpk8ZK868U/iXBTtdpD3bBhg/bu3Ws/Yegyd7z9IHuoAK4k7KEay6l7qDNmzFBaWpq6devG0+ABAPhJjQv1448/1vr16+03WQAAALW4bCYiIoJpBAAAfqXGe6hz587V7NmzdfvttyssrOzN3Tt06GBYMAAA6pMaF2pWVpY+//xzffDBB2WWWywWZWRkGBYMAID6pMaF+tRTT2n69OkaOnQoU78AAPykxoVaXFyskSNHysvL7e9aCACAy9S4FceMGaOVK1c6IwsAwF2UFEmZH0rpKaX/W1JkdiK3V+M91I0bN2r//v1atGiRrr667H0u165da1gwAIBJdv1HSk2ScrJ+XhYUId38hNTlTsM317591U+4uf/++/XAAw8Yvl2j1bhQ77nnHmfkAAC4g13/kdZNKr88J+vn5QaX6ubNm+1fv//++3rhhRf04Ycf2pddfuqYVPpA8JKSEvn4uN9tImucaMSIEc7IAQAwW0lR6Z5pVVKTpN/+wdD7HoeH//zA8uDgYFksFvuyrVu36u6779aSJUuUnJysb7/9VkuXLtXatWt18eJFvfTSS/b3zps3T/v27dOKFSskSVarVa+88opWrVqls2fPqnXr1po0aZIGDx5sWPZfcugnsmnTJt14442SpNTU1ErH3XzzzcakAgC43oHUstO8FcnJkg6mSu0GuSbTTxYuXKhHH31ULVq0UEhIiEPvWbx4sd555x3NnTtXrVu31vbt2zVt2jSFhYWpe/fuhmd0qFAXLFhgL9R58+ZVOMZisVCoAFCfVVeml2U7OM5AkydPVu/evR0eX1hYqMWLF+vVV19Vly5dJJU++zQ9PV2rVq0yr1Dfe+89+9cbN240PAQAwA0ERVQ/RpKCHRxnoE6dOtVo/NGjR3Xp0iWNHz++zPKioiJFR0cbGc3O/Y7qAgDMce3NpaVa1Z5qUITU1vWzkb9+upnFYtGvnz5aXFxs//ry40UXL16spk2blhnn5+fnlIwOFWq/fv1ksViqHVfV8VUAgJvz9i29NKais3wv6z/bLR7EHhYWpv3795dZlpGRIV9fX0lS27Zt5efnp5MnTzplerciDv1UHn/8cfvXBw8e1BtvvKE77rhDzZs314kTJ/TGG29o9OjRTgsJAHCRy5fEVHQdav/ZUswYc3L9So8ePbR06VK9/fbbiomJ0TvvvKP9+/erY8eOkqSgoCCNHz9e8+fPl81mU2xsrLKzs7Vz504FBQU55YoVhwq1f//+9q9ffvll/fOf/1SbNm3sy26++WZNmzZNEyZMMDwgAMDFutxZemnMwdTSE5CCf5rmdYM908v69u2rSZMmacGCBSooKNCoUaM0fPhwffvtt/YxU6dOVVhYmBYvXqzjx48rODhYHTt21MSJE52SyWL79SR0NWJjY/Xll1+WmYMuKChQr169lJ6ebnjAusrJyVFsbKzS09MVFBRkdhwAcKr8/HwdPnxYbdq04QEmBqjJz7PG9/KNiYnR7Nmz9eOPP0qSzp07p6SkJHXu3Ll2aQEA8AA1LtSnnnpK3333nXr16qWYmBj17t1bR44c0VNPPeWMfAAA1As1nhBv2rSp/vOf/ygrK0unT59WkyZNFBHh+muSAABwJ7U+whwREUGRAgDwE54SDgAeqIbnm6ISNfk5UqgA4EG8vb0lld7LFnV3+ed4+edaFfe5qAgAUGc+Pj4KCAjQmTNn5OvrKy8v9ptqy2q16syZMwoICHDo+asUKgB4EIvFosjISB0+fFhHjx41O0695+XlpZYtWzp0+10KFQA8jJ+fn6677jqmfQ3g5+fn8F4+hQoAHsjLy4s7JbkYk+sAABiAQgUAwAAUKgAABqBQAQAwAIUKAIABKFQAAAxAoQIAYAAKFQAAA1CoAAAYwK0LNS0tTXfeeae6deumnj17avLkycrKyjI7FgAA5bh1oWZnZysxMVFpaWlKTU1VYGCgpk6danYsAADKcet7+SYkJJR5fc8992jEiBEqLi526FE6AAC4ilvvof7a9u3b1bZtW8oUAOB2TGumCRMmKC0trdL1qampioqKsr/+5ptvlJycrOTkZBekAwCgZkwr1IULF1b5rL5GjRrZv87MzFRiYqJmzZql3r17uyAdAAA1Y1qhBgUFOTQuMzNT48aN00MPPaRhw4Y5ORUAALXj1sdQ9+/fr3Hjxmnq1KkaNWqU2XEAAKiUWxfqsmXLdO7cOc2fP19dunSx/zl58qTZ0QAAKMNis9lsZodwppycHMXGxio9Pd3haWYAAGrKrfdQAQCoLyhUAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAEoVAAADEChAgBgAAoVAAADUKgAABiAQgUAwAAUKgAABqBQAQAwgI/ZATxKdpZ0+hvJx1+KipO8fc1OBABwEQrVCBdOSB8+KmV+IFmLS5cFNZV6TJJ6T5EsFnPzAQCcjkKtq+xT0rJB0oVjZZfnnJI2zJYunpBuXWBONgCAy3AMta4+X1i+TH9p2xLpdIbr8gAATEGh1kVJkfTVyurH7fyX87MAAExFodbFpR+lgovVj/vxiNOjAADMRaHWhV+Q5OXAYWj/MOdnAQCYikKtC78Aqf2t1Y/rNNr5WQAApqJQ6+qGhyWfhpWvb9VbuuZ3LosDADAHhVpXkZ2lMW9IwZHl17W7RfrjSq5DBYArANehGuGaG6Wpe6TM96VTeyWfBlKHIVJ4e7OTAQBchEI1ireP1PG20j8AgCsOU74AABiAQgUAwAAUKgAABqBQAQAwAIUKAIABKFQAAAxAoQIAYAAKFQAAA1CoAAAYgEIFAMAAFCoAAAaoN4W6atUqtW/fXikpKWZHAQCgnHpRqKdOndLSpUvVrl07s6MAAFChelGoSUlJmjRpkho1amR2FAAAKuT2hfrhhx8qJydHw4cPNzsKAACVMu15qBMmTFBaWlql61NTUxUcHKxnn31Wy5Ytc10wAABqwbRCXbhwoQoLCytd36hRI82aNUujR49W69atXRcMAIBaMK1Qg4KCqh3z5ZdfKicnR8uXL5ck5eTkaM+ePUpPT9eLL77o7IgAADjMtEJ1xKpVq1RSUmJ/PWXKFPXt21djxowxMRUAAOW5daGGh4eXee3n56egoCCFhYWZlAgAgIq5daH+2ooVK8yOAABAhdz+shkAAOoDChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAEoVAAADFCvbuwAx13ML9IXB35QYYlVv2kWorbh1d87GQBQexSqhykqsWr++/v0+vbvlFf4832Qe197leaP+K1aXhVgYjoA8FxM+XqYySt3admWw2XKVJK2HPhBty/+QlkX8k1KBgCejUL1INsOn9MHe7IqXX/qYoEWbTrowkQAcOWgUD3ImzuOVTtmzc7jslptLkgDAFcWCtWDZF2sfjo3O79YuYXFLkgDAFcWCtWDXB3UoNox3l4WBfhxLhoAGI1C9SAjujSvdkyJ1aYdR865IA0AXFkoVA/S97qr5edd/f+lr2+v/lgrAKBmKFQPYrVJhSXWasd9dy7PBWkA4MpCoXoQby+LghpUf3y0cYCvC9IAwJWFQvUwt8U0c2BM9cdaAQA1Q6F6mHv7XqPghpXvpUZHhuiW6yNcmAgArgwUqodpfXWgVvw5Xi3C/Mut63FNmP41vrt8HThxCQBQM1yQ6IFiWjTSpodv0qeZp/XV8Qvy9bLopg5NdH3zULOjAYDHolA9lJeXRTdHN9XN0U3NjgIAVwTm/gAAMACFCgCAAShUAAAMwDFUAHA1q1U6vEm6eFIKvFpq20/y5oYr9R2FCgCulPGu9NHj0vnvfl4W2ES66TEpbrx5uVBnFCoAuMq+9dIbd0u2X91zO/e09N7fpJJiKf5ec7KhzjiGCgCuYLNJnzxRvkx/6dP/kQp5eEV9RaECgCsc2yr9cKDqMfkXpMz3XZMHhqNQAcAVsr93bNzFk87NAaehUAHAFYIcvGtZMA+vqK8oVABwhZY9pcZtqh7TIETqMMQ1eWA4ChUAXMFikfrPkWSpfMyNj0p+ga5KBINRqADgKr8ZLo36pxTcrOxy/zBp8DNSr/tNiQVjcB0qALhSp9FSx+HSwY3ShWNSYLh03UDJt6HZyVBHFCoAuJq3j9RuoNkpYDC3n/K9ePGiZsyYofj4eHXt2lUjR47UpUuXzI4FAEAZbr2HarVaNWHCBLVr104fffSRQkJCtG/fPvn4ODl2QU7pVIxvgNS4lXO3BQDwCG5dqJ999pm+//57rVixwl6iHTt2dN4Gc86U3vpr95tSUW7pssgYqe+DUsdhztsuAKDec+sp323btqlly5Z65JFHFB8fryFDhmjt2rXO2VjuWWnZICk95ecylaTv/1t6M+uti52zXQCARzBtD3XChAlKS0urdH1qaqouXLigrVu3atasWXr66af19ddf6y9/+YuioqLUrVs3YwN9+pR07mDl6z+aIf1mhBTUxNjtAgA8gmmFunDhQhUWFla6vlGjRgoICFBERITuuusuSVJsbKz69++vTz/91NhCLcyTdr9R9RhrkbRrhdT3IeO2CwDwGKYValBQULVjOnTooI8//tj5YS6elAqzqx93dr/zswAA6iW3PoY6YMAAFRQUaOXKlSopKdFXX32l1NRU9evXz9gNNai+3CVxSzAAQKXculBDQkK0ZMkSrV69WrGxsXrkkUf0xBNPKC4uztgNBUdILXpUP67jcGO3CwDwGG592Ywk/fa3v9WaNWucv6G+D0mv/V6SreL1LXpIbfo6PwcAoF5y6z1Ul2o3UEpIlnwquJ9mix7SHa+5PhMAoN5w+z1Ul4q9R4pOkL5aKZ3OkPyCpI63Sa16mZ0MAODmKNRfCwiTet5ndgoAQD3DlC8AAAagUAEAMACFCgCAAShUAAAMQKECAGAAChUAAANQqAAAGMDjr0O12UpvJZiTk2NyEgAwVmBgoCwWi9kx8BOPL9Tc3FxJ0o033mhyEgAwVnp6ukOPwoRrWGyXd+E8lNVq1enTp/lNDoDH4d819+LxhQoAgCtwUhIAAAagUAEAMACFCgCAAShUAAAMQKECAGAAChUAAANQqAAAGIBCBQDAAFdkoa5atUrt27dXSkqK2VGcJi0tTXfeeae6deumnj17avLkycrKyjI7lmGKioqUlJSkbt26qXv37nryySdVXFxsdizDFRYWaubMmerXr5+6dOmiwYMHa/Xq1WbHcon8/HwNGDBAcXFxZkdxqtTUVA0bNkwxMTHq06ePVq5caXYk1NIVV6inTp3S0qVL1a5dO7OjOFV2drYSExOVlpam1NRUBQYGaurUqWbHMszLL7+s9PR0rV+/Xu+995527NihRYsWmR3LcMXFxQoPD1dKSop27typp59+Ws8884w2b95sdjSnS05OVrNmzcyO4VSfffaZ5s6dq8cff9z+97l79+5mx0ItXXGFmpSUpEmTJqlRo0ZmR3GqhIQE/e53v1NgYKACAgJ0zz336KuvvvKYvbg1a9bor3/9q5o0aaImTZpo4sSJWrNmjdmxDBcQEKApU6aoZcuWslgsiomJUXx8vNLT082O5lR79uzR5s2blZiYaHYUp0pOTtZ9992n+Ph4eXt7KzQ0VG3btjU7FmrpiirUDz/8UDk5ORo+fLjZUVxu+/btatu2rXx86v8Dhi5cuKCsrCxFR0fbl0VHR+vkyZPKzs42MZnzFRQUaPfu3Wrfvr3ZUZymuLhYs2bN0hNPPCFfX1+z4zhNXl6e9u7dq1OnTmnQoEHq3bu3Jk+erNOnT5sdDbVU//91lTRhwgSlpaVVuj41NVXBwcF69tlntWzZMtcFcxJHPm9UVJT99TfffKPk5GQlJye7IJ3z5eXlSZKCg4Pty0JCQiSVPq7vl8s9ic1m04wZM9SqVSsNHDjQ7DhOs3TpUkVHR6tbt27aunWr2XGc5uLFi7LZbNqwYYOWLVumRo0aafbs2Zo2bZqWL19udjzUgkcU6sKFC1VYWFjp+kaNGmnWrFkaPXq0Wrdu7bpgTuLI570sMzNTiYmJmjVrlnr37u2CdM4XEBAgqfSh8WFhYZJk3zMNDAw0LZcz2Ww2zZkzR4cPH1ZKSoq8vDxzcuno0aN6/fXXtXbtWrOjON3lv8djx45V8+bNJUmTJ0/WwIEDlZeXZ1+P+sMjCtWRB+x++eWXysnJsf/ml5OToz179ig9PV0vvviisyMaytEHCmdmZmrcuHF66KGHNGzYMCencp3Q0FBFREQoIyNDLVu2lCRlZGQoMjLSI/dObTab5s6dq927dyslJcUjP+Nl6enpOnv2rAYNGiSpdPo3NzdX8fHxWrJkiTp37mxyQuOEhIRUetIVT9WsnzyiUB2xatUqlZSU2F9PmTJFffv21ZgxY0xM5Tz79+/XuHHjNHXqVI0aNcrsOIYbOXKkFi1apK5du0qSFi9erNGjR5ucyjmSkpK0c+dOLV++XKGhoWbHcapbbrlFvXr1sr/etWuXZs6cqXXr1tlnIzzJ73//e/373//WDTfcoNDQUP3jH/9Qz549PXamxdNdMYUaHh5e5rWfn5+CgoI88j9SSVq2bJnOnTun+fPna/78+fbl69ev94hLESZNmqTz58/r1ltvlSTddtttmjhxosmpjHfixAm99tpr8vPzU79+/ezLExISlJSUZGIy5/D395e/v7/9dVhYmCwWiyIiIkxM5Tz33nuvLly4oNtuu02SFB8fr2effdbkVKgti425BQAA6swzz2wAAMDFKFQAAAxAoQIAYAAKFQAAA1CoAAAYgEIFAMAAFCoAAAagUAEAMACFClRiw4YNZe5ONGTIEH366acmJgLgzihUwEHr16/XTTfdZNj3O3jwoO644w517txZgwYNUmpqqmHfG4DrUaiACYqKijRx4kT17NlT27Zt0/Tp0/Xwww/r6NGjZkcDUEsUKvCTrKwsjR8/Xl27dtXIkSN14MCBMuv79eunDRs2SJLeeustDRs2TMnJyYqPj1fv3r31/vvvKz09XUOHDlVsbKwef/xxWa3WCre1Y8cOnT9/XpMmTVKDBg100003qVu3blq3bp3TPycA57hinjYDVOehhx5SVFSUtmzZopMnTyoxMbHK8fv379eoUaO0ZcsWrV27VrNmzVKfPn20YsUKFRYWasSIEdqwYYMGDhxY7r2ZmZm69tpr5evra18WHR2tzMxMwz8XANdgDxWQ9P3332vHjh165JFH5O/vr7Zt2+qOO+6o8j1hYWG6++675ePjo6FDhyonJ0ejR49W48aN1bRpU3Xr1k179+6t8L25ubkKCQkpsyw4OFi5ubmGfSYArkWhApJOnz6tBg0a6KqrrrIva968eZXv+eXYy8/wvPrqq8ssy8vLq/C9gYGBys7OLrMsOzubB0sD9RiFCkhq0qSJCgoK9MMPP9iXnTx50mnba9++vQ4cOKCioiL7sn379qldu3ZO2yYA56JQAUmRkZHq2rWr/v73vys/P1+HDh3SqlWrnLa9uLg4hYaGatGiRSosLNSmTZu0bds2DR8+3GnbBOBcFCrwk4ULFyorK0s9e/bUww8/rFGjRjltW76+vnr55Ze1ZcsWxcXFad68eVqwYIFatWrltG0CcC6LzWazmR0CAID6jj1UAAAMQKECAGAAChUAAANQqAAAGIBCBQDAABQqAAAGoFABADAAhQoAgAEoVAAADEChAgBgAAoVAAAD/H8U6tkLKMOLKQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 400x400 with 1 Axes>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_condition_embedding(df_condition_embedding, embedding=\"PCA\", hue=\"seen_during_training\", circle_size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "619785af-e6e0-473d-bf9f-ee10984761fb",
   "metadata": {},
   "source": [
    "## Predicting with CellFlow\n",
    "\n",
    "Predictions with {class}`~cellflow.model.CellFlow` require an {class}`adata <anndata.AnnData>` object with control cells. Moreover, we need `covariate_data` to store the information about what we would like to predict. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c5709fd8-4675-4805-ac90-25f2c5ead19b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>n.umi</th>\n",
       "      <th>RT_well</th>\n",
       "      <th>Drug1</th>\n",
       "      <th>Drug2</th>\n",
       "      <th>Well</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>control</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>is_control</th>\n",
       "      <th>Drug1_pubchem_name</th>\n",
       "      <th>Drug1_pubchem_ID</th>\n",
       "      <th>Drug1_smiles</th>\n",
       "      <th>Drug2_pubchem_name</th>\n",
       "      <th>Drug2_pubchem_ID</th>\n",
       "      <th>Drug2_smiles</th>\n",
       "      <th>mode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_22_Lig_BC_17</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.899999</td>\n",
       "      <td>2416</td>\n",
       "      <td>RT_22</td>\n",
       "      <td>control</td>\n",
       "      <td>Dacinostat</td>\n",
       "      <td>B10</td>\n",
       "      <td>1572</td>\n",
       "      <td>1569</td>\n",
       "      <td>2412.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dacinostat</td>\n",
       "      <td>6445533</td>\n",
       "      <td>C1=CC=C2C(=C1)C(=CN2)CCN(CCO)CC3=CC=C(C=C3)C=C...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_28_Lig_BC_12</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>1.136175</td>\n",
       "      <td>3050</td>\n",
       "      <td>RT_28</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>Sorafenib</td>\n",
       "      <td>C4</td>\n",
       "      <td>2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>3050.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "      <td>Sorafenib</td>\n",
       "      <td>216239</td>\n",
       "      <td>CNC(=O)C1=NC=CC(=C1)OC2=CC=C(C=C2)NC(=O)NC3=CC...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_4_Lig_BC_15</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.932036</td>\n",
       "      <td>2502</td>\n",
       "      <td>RT_4</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>SRT1720</td>\n",
       "      <td>A4</td>\n",
       "      <td>1649</td>\n",
       "      <td>1649</td>\n",
       "      <td>2502.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "      <td>1001645-58-4</td>\n",
       "      <td>25232708</td>\n",
       "      <td>C1CN(CCN1)CC2=CSC3=NC(=CN23)C4=CC=CC=C4NC(=O)C...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_52_Lig_BC_32</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>2.259311</td>\n",
       "      <td>6065</td>\n",
       "      <td>RT_52</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>Crizotinib</td>\n",
       "      <td>E4</td>\n",
       "      <td>3262</td>\n",
       "      <td>3260</td>\n",
       "      <td>6063.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "      <td>Crizotinib</td>\n",
       "      <td>11626560</td>\n",
       "      <td>CC(C1=C(C=CC(=C1Cl)F)Cl)OC2=C(N=CC(=C2)C3=CN(N...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_70_Lig_BC_12</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>1.894990</td>\n",
       "      <td>5087</td>\n",
       "      <td>RT_70</td>\n",
       "      <td>control</td>\n",
       "      <td>Alvespimycin</td>\n",
       "      <td>F10</td>\n",
       "      <td>2801</td>\n",
       "      <td>2800</td>\n",
       "      <td>5086.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>control</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>alvespimycin</td>\n",
       "      <td>5288674</td>\n",
       "      <td>CC1CC(C(C(C=C(C(C(C=CC=C(C(=O)NC2=CC(=O)C(=C(C...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_73_Lig_BC_11</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.539404</td>\n",
       "      <td>1448</td>\n",
       "      <td>RT_73</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>Curcumin</td>\n",
       "      <td>G1</td>\n",
       "      <td>1114</td>\n",
       "      <td>1113</td>\n",
       "      <td>1447.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>6918837</td>\n",
       "      <td>CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO</td>\n",
       "      <td>curcumin</td>\n",
       "      <td>969516</td>\n",
       "      <td>COC1=C(C=CC(=C1)C=CC(=O)CC(=O)C=CC2=CC(=C(C=C2...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A01_A02_RT_BC_79_Lig_BC_57</th>\n",
       "      <td>sciPlex_theis</td>\n",
       "      <td>0.772226</td>\n",
       "      <td>2073</td>\n",
       "      <td>RT_79</td>\n",
       "      <td>SRT2104</td>\n",
       "      <td>Alvespimycin</td>\n",
       "      <td>G7</td>\n",
       "      <td>1459</td>\n",
       "      <td>1456</td>\n",
       "      <td>2070.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A549</td>\n",
       "      <td>False</td>\n",
       "      <td>1093403-33-8</td>\n",
       "      <td>25108829</td>\n",
       "      <td>CC1=C(SC(=N1)C2=CN=CC=C2)C(=O)NC3=CC=CC=C3C4=C...</td>\n",
       "      <td>alvespimycin</td>\n",
       "      <td>5288674</td>\n",
       "      <td>CC1CC(C(C(C=C(C(C(C=CC=C(C(=O)NC2=CC(=O)C(=C(C...</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   sample  Size_Factor  n.umi RT_well  \\\n",
       "Cell                                                                    \n",
       "A01_A02_RT_BC_22_Lig_BC_17  sciPlex_theis     0.899999   2416   RT_22   \n",
       "A01_A02_RT_BC_28_Lig_BC_12  sciPlex_theis     1.136175   3050   RT_28   \n",
       "A01_A02_RT_BC_4_Lig_BC_15   sciPlex_theis     0.932036   2502    RT_4   \n",
       "A01_A02_RT_BC_52_Lig_BC_32  sciPlex_theis     2.259311   6065   RT_52   \n",
       "A01_A02_RT_BC_70_Lig_BC_12  sciPlex_theis     1.894990   5087   RT_70   \n",
       "A01_A02_RT_BC_73_Lig_BC_11  sciPlex_theis     0.539404   1448   RT_73   \n",
       "A01_A02_RT_BC_79_Lig_BC_57  sciPlex_theis     0.772226   2073   RT_79   \n",
       "\n",
       "                                   Drug1         Drug2 Well  n_genes  \\\n",
       "Cell                                                                   \n",
       "A01_A02_RT_BC_22_Lig_BC_17       control    Dacinostat  B10     1572   \n",
       "A01_A02_RT_BC_28_Lig_BC_12  Panobinostat     Sorafenib   C4     2015   \n",
       "A01_A02_RT_BC_4_Lig_BC_15   Panobinostat       SRT1720   A4     1649   \n",
       "A01_A02_RT_BC_52_Lig_BC_32  Panobinostat    Crizotinib   E4     3262   \n",
       "A01_A02_RT_BC_70_Lig_BC_12       control  Alvespimycin  F10     2801   \n",
       "A01_A02_RT_BC_73_Lig_BC_11  Panobinostat      Curcumin   G1     1114   \n",
       "A01_A02_RT_BC_79_Lig_BC_57       SRT2104  Alvespimycin   G7     1459   \n",
       "\n",
       "                            n_genes_by_counts  total_counts  ...  control  \\\n",
       "Cell                                                         ...            \n",
       "A01_A02_RT_BC_22_Lig_BC_17               1569        2412.0  ...        0   \n",
       "A01_A02_RT_BC_28_Lig_BC_12               2015        3050.0  ...        0   \n",
       "A01_A02_RT_BC_4_Lig_BC_15                1649        2502.0  ...        0   \n",
       "A01_A02_RT_BC_52_Lig_BC_32               3260        6063.0  ...        0   \n",
       "A01_A02_RT_BC_70_Lig_BC_12               2800        5086.0  ...        0   \n",
       "A01_A02_RT_BC_73_Lig_BC_11               1113        1447.0  ...        0   \n",
       "A01_A02_RT_BC_79_Lig_BC_57               1456        2070.0  ...        0   \n",
       "\n",
       "                            cell_type is_control Drug1_pubchem_name  \\\n",
       "Cell                                                                  \n",
       "A01_A02_RT_BC_22_Lig_BC_17       A549      False            control   \n",
       "A01_A02_RT_BC_28_Lig_BC_12       A549      False       Panobinostat   \n",
       "A01_A02_RT_BC_4_Lig_BC_15        A549      False       Panobinostat   \n",
       "A01_A02_RT_BC_52_Lig_BC_32       A549      False       Panobinostat   \n",
       "A01_A02_RT_BC_70_Lig_BC_12       A549      False            control   \n",
       "A01_A02_RT_BC_73_Lig_BC_11       A549      False       Panobinostat   \n",
       "A01_A02_RT_BC_79_Lig_BC_57       A549      False       1093403-33-8   \n",
       "\n",
       "                           Drug1_pubchem_ID  \\\n",
       "Cell                                          \n",
       "A01_A02_RT_BC_22_Lig_BC_17             <NA>   \n",
       "A01_A02_RT_BC_28_Lig_BC_12          6918837   \n",
       "A01_A02_RT_BC_4_Lig_BC_15           6918837   \n",
       "A01_A02_RT_BC_52_Lig_BC_32          6918837   \n",
       "A01_A02_RT_BC_70_Lig_BC_12             <NA>   \n",
       "A01_A02_RT_BC_73_Lig_BC_11          6918837   \n",
       "A01_A02_RT_BC_79_Lig_BC_57         25108829   \n",
       "\n",
       "                                                                 Drug1_smiles  \\\n",
       "Cell                                                                            \n",
       "A01_A02_RT_BC_22_Lig_BC_17                                                NaN   \n",
       "A01_A02_RT_BC_28_Lig_BC_12    CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO   \n",
       "A01_A02_RT_BC_4_Lig_BC_15     CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO   \n",
       "A01_A02_RT_BC_52_Lig_BC_32    CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO   \n",
       "A01_A02_RT_BC_70_Lig_BC_12                                                NaN   \n",
       "A01_A02_RT_BC_73_Lig_BC_11    CC1=C(C2=CC=CC=C2N1)CCNCC3=CC=C(C=C3)C=CC(=O)NO   \n",
       "A01_A02_RT_BC_79_Lig_BC_57  CC1=C(SC(=N1)C2=CN=CC=C2)C(=O)NC3=CC=CC=C3C4=C...   \n",
       "\n",
       "                           Drug2_pubchem_name  Drug2_pubchem_ID  \\\n",
       "Cell                                                              \n",
       "A01_A02_RT_BC_22_Lig_BC_17         Dacinostat           6445533   \n",
       "A01_A02_RT_BC_28_Lig_BC_12          Sorafenib            216239   \n",
       "A01_A02_RT_BC_4_Lig_BC_15        1001645-58-4          25232708   \n",
       "A01_A02_RT_BC_52_Lig_BC_32         Crizotinib          11626560   \n",
       "A01_A02_RT_BC_70_Lig_BC_12       alvespimycin           5288674   \n",
       "A01_A02_RT_BC_73_Lig_BC_11           curcumin            969516   \n",
       "A01_A02_RT_BC_79_Lig_BC_57       alvespimycin           5288674   \n",
       "\n",
       "                                                                 Drug2_smiles  \\\n",
       "Cell                                                                            \n",
       "A01_A02_RT_BC_22_Lig_BC_17  C1=CC=C2C(=C1)C(=CN2)CCN(CCO)CC3=CC=C(C=C3)C=C...   \n",
       "A01_A02_RT_BC_28_Lig_BC_12  CNC(=O)C1=NC=CC(=C1)OC2=CC=C(C=C2)NC(=O)NC3=CC...   \n",
       "A01_A02_RT_BC_4_Lig_BC_15   C1CN(CCN1)CC2=CSC3=NC(=CN23)C4=CC=CC=C4NC(=O)C...   \n",
       "A01_A02_RT_BC_52_Lig_BC_32  CC(C1=C(C=CC(=C1Cl)F)Cl)OC2=C(N=CC(=C2)C3=CN(N...   \n",
       "A01_A02_RT_BC_70_Lig_BC_12  CC1CC(C(C(C=C(C(C(C=CC=C(C(=O)NC2=CC(=O)C(=C(C...   \n",
       "A01_A02_RT_BC_73_Lig_BC_11  COC1=C(C=CC(=C1)C=CC(=O)CC(=O)C=CC2=CC(=C(C=C2...   \n",
       "A01_A02_RT_BC_79_Lig_BC_57  CC1CC(C(C(C=C(C(C(C=CC=C(C(=O)NC2=CC(=O)C(=C(C...   \n",
       "\n",
       "                            mode  \n",
       "Cell                              \n",
       "A01_A02_RT_BC_22_Lig_BC_17  test  \n",
       "A01_A02_RT_BC_28_Lig_BC_12  test  \n",
       "A01_A02_RT_BC_4_Lig_BC_15   test  \n",
       "A01_A02_RT_BC_52_Lig_BC_32  test  \n",
       "A01_A02_RT_BC_70_Lig_BC_12  test  \n",
       "A01_A02_RT_BC_73_Lig_BC_11  test  \n",
       "A01_A02_RT_BC_79_Lig_BC_57  test  \n",
       "\n",
       "[7 rows x 27 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ctrl_for_prediction = adata_test[adata_test.obs[\"is_control\"].to_numpy()]\n",
    "covariate_data = adata_test[~adata_test.obs[\"is_control\"].to_numpy()].obs.drop_duplicates(subset=[\"condition\"])\n",
    "covariate_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b1933f-61dc-4019-8363-bbc259539048",
   "metadata": {},
   "source": [
    "We specify `sample_rep` to be `\"X_pca\"`, and provide a `condition_id_key` which defines the keys of the returned dictionary containing the predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "af28310d-825c-42b3-b50f-e8c3452e9eb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7/7 [00:00<00:00, 1006.76it/s]\n"
     ]
    }
   ],
   "source": [
    "preds = cf.predict(adata=adata_ctrl_for_prediction, sample_rep=\"X_aligned\", condition_id_key=\"condition\", covariate_data=covariate_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936de826-30b1-4ff5-8770-ec2285e70712",
   "metadata": {},
   "source": [
    "We now build an {class}`adata <anndata.AnnData>` object to store the predictions in {attr}`adata.obsm <anndata.AnnData.obsm>`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "28534185-7ecc-4783-b30c-32adaafd9498",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_preds = []\n",
    "for cond, array in preds.items():\n",
    "    \n",
    "    obs_data = pd.DataFrame({\n",
    "        'condition': [cond] * array.shape[0]\n",
    "    })\n",
    "    adata_pred = ad.AnnData(X=np.empty((len(array),adata_train.n_vars)), obs=obs_data)\n",
    "    adata_pred.obsm[\"X_pca\"] = np.squeeze(array)\n",
    "    adata_preds.append(adata_pred)\n",
    "\n",
    "adata_preds = ad.concat(adata_preds)\n",
    "adata_preds.var_names = adata_train.var_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d6cfef-4e71-4dff-a545-ba6244d63f2d",
   "metadata": {},
   "source": [
    "We now reconstruct the generated cells to gene space, which adds {attr}`uns['X_recon'] <anndata.AnnData.layers>` added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6c959d20-348f-476d-a592-f74711a4fd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "reconstruct_pca(adata_preds, use_rep=\"X_pca\", ref_adata=adata_train)\n",
    "adata_preds.X = adata_preds.layers[\"X_recon\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b290bcf-30c7-4077-b3ac-959dabed0aee",
   "metadata": {},
   "source": [
    "We also compute the metrics of CellFlow with respect to the ground truth data going through the encoder-decoder in order to separate CellFlow's model performance from the encoder-decoder. Note that this is what is computed during training with {class}`~cellflow.training.PCADecodedMetrics`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "011b9096-a8f0-4a6b-8c8f-f60446d7eef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_test_recon = adata_test.copy()\n",
    "cfpp.project_pca(adata_test_recon, ref_adata=adata_train)\n",
    "cfpp.reconstruct_pca(query_adata=adata_test_recon, ref_adata=adata_train, layers_key_added=\"X_recon\")\n",
    "adata_test_recon.X = adata_test_recon.layers[\"X_recon\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd3fd70e-c4c9-452c-9af8-c7d2faa74ade",
   "metadata": {},
   "source": [
    "We now generate a latent space based on the whole data for evaluation of distributional metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "224e02fd-d5a8-4cb5-99fa-7cd235605047",
   "metadata": {},
   "outputs": [],
   "source": [
    "cfpp.centered_pca(adata, n_comps=30, method=\"rapids\")\n",
    "cfpp.project_pca(query_adata=adata_test, ref_adata=adata)\n",
    "cfpp.project_pca(query_adata=adata_preds, ref_adata=adata)\n",
    "cfpp.project_pca(query_adata=adata_test_recon, ref_adata=adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d8eaa1ee-a6ba-48d5-9260-5148a2eeafa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data_target_encoded = {}\n",
    "test_data_target_decoded = {}\n",
    "test_data_target_encoded_reconstructed = {}\n",
    "test_data_target_decoded_reconstructed = {}\n",
    "test_data_target_encoded_predicted = {}\n",
    "test_data_target_decoded_predicted = {}\n",
    "\n",
    "for cond in adata_preds.obs[\"condition\"].unique():\n",
    "    test_data_target_encoded[cond] = adata_test[adata_test.obs[\"condition\"] == cond].obsm[\"X_pca\"]\n",
    "    test_data_target_decoded[cond] = adata_test[adata_test.obs[\"condition\"] == cond].X.toarray()\n",
    "    test_data_target_encoded_reconstructed[cond] = adata_test_recon[adata_test_recon.obs[\"condition\"] == cond].obsm[\"X_pca\"]\n",
    "    test_data_target_decoded_reconstructed[cond] = adata_test_recon[adata_test_recon.obs[\"condition\"] == cond].X.toarray()\n",
    "    test_data_target_encoded_predicted[cond] = adata_preds[adata_preds.obs[\"condition\"] == cond].obsm[\"X_pca\"]\n",
    "    test_data_target_decoded_predicted[cond] = adata_preds[adata_preds.obs[\"condition\"] == cond].X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8fc90cd2-c9ba-4e4f-b039-a6d7200d5a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_distances_true_pred = jax.tree_util.tree_map(\n",
    "    compute_e_distance, test_data_target_encoded, test_data_target_encoded_predicted\n",
    ")\n",
    "\n",
    "r_squared_true_pred = jax.tree_util.tree_map(\n",
    "    compute_r_squared, test_data_target_decoded, test_data_target_decoded_predicted\n",
    ")\n",
    "\n",
    "df_e_distance_true_pred = pd.DataFrame.from_dict(e_distances_true_pred, orient=\"index\", columns=[\"energy_distance\"])\n",
    "df_r_squared_true_pred = pd.DataFrame.from_dict(r_squared_true_pred, orient=\"index\", columns=[\"r_squared\"])\n",
    "df_metrics_true_pred = pd.merge(df_e_distance_true_pred, df_r_squared_true_pred, left_index=True, right_index=True)\n",
    "df_metrics_true_pred[\"condition\"] = df_metrics_true_pred.index\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14, 5), sharex=True)\n",
    "\n",
    "sns.barplot(\n",
    "    data=df_metrics_true_pred,\n",
    "    x=\"condition\",\n",
    "    y=\"energy_distance\",\n",
    "    edgecolor=\"black\",\n",
    "    ax=axes[0]\n",
    ")\n",
    "axes[0].set_title(\"Energy Distance in latent space\", fontsize=14)\n",
    "axes[0].set_xlabel(\"Drug treatment\", fontsize=12)\n",
    "axes[0].set_ylabel(\"Energy Distance\", fontsize=12)\n",
    "axes[0].tick_params(axis='x', rotation=45)\n",
    "\n",
    "sns.barplot(\n",
    "    data=df_metrics_true_pred,\n",
    "    x=\"condition\",\n",
    "    y=\"r_squared\",\n",
    "    edgecolor=\"black\",\n",
    "    ax=axes[1]\n",
    ")\n",
    "axes[1].set_title(\"R squared in Gene Expression Space\", fontsize=14)\n",
    "axes[1].set_xlabel(\"Drug treatment\", fontsize=12)\n",
    "axes[1].set_ylabel(\"R squared\", fontsize=12)\n",
    "axes[1].tick_params(axis='x', rotation=45)\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "09eb9d4c-a4b0-4d75-9c99-3b0bf65baa1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_distances_recon_pred = jax.tree_util.tree_map(\n",
    "    compute_e_distance, test_data_target_encoded_reconstructed, test_data_target_encoded_predicted\n",
    ")\n",
    "\n",
    "r_squared_recon_pred = jax.tree_util.tree_map(\n",
    "    compute_r_squared, test_data_target_decoded_reconstructed, test_data_target_decoded_predicted\n",
    ")\n",
    "\n",
    "df_e_distance_recon_pred = pd.DataFrame.from_dict(e_distances_recon_pred, orient=\"index\", columns=[\"energy_distance\"])\n",
    "df_r_squared_recon_pred = pd.DataFrame.from_dict(r_squared_recon_pred, orient=\"index\", columns=[\"r_squared\"])\n",
    "df_metrics_recon_pred = pd.merge(df_e_distance_recon_pred, df_r_squared_recon_pred, left_index=True, right_index=True)\n",
    "df_metrics_recon_pred[\"condition\"] = df_metrics_recon_pred.index\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14, 5), sharex=True)\n",
    "\n",
    "sns.barplot(\n",
    "    data=df_metrics_recon_pred,\n",
    "    x=\"condition\",\n",
    "    y=\"energy_distance\",\n",
    "    edgecolor=\"black\",\n",
    "    ax=axes[0]\n",
    ")\n",
    "axes[0].set_title(\"Energy Distance in latent space\", fontsize=14)\n",
    "axes[0].set_xlabel(\"Drug treatment\", fontsize=12)\n",
    "axes[0].set_ylabel(\"Energy Distance\", fontsize=12)\n",
    "axes[0].tick_params(axis='x', rotation=45)\n",
    "\n",
    "sns.barplot(\n",
    "    data=df_metrics_recon_pred,\n",
    "    x=\"condition\",\n",
    "    y=\"r_squared\",\n",
    "    edgecolor=\"black\",\n",
    "    ax=axes[1]\n",
    ")\n",
    "axes[1].set_title(\"R squared in Gene Expression Space\", fontsize=14)\n",
    "axes[1].set_xlabel(\"Drug treatment\", fontsize=12)\n",
    "axes[1].set_ylabel(\"R squared\", fontsize=12)\n",
    "axes[1].tick_params(axis='x', rotation=45)\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d911d590-663e-40e9-8942-325207fd6b91",
   "metadata": {},
   "source": [
    "From the comparison of the plots we can conclude the following:\n",
    "- CellFlow has almost perfectly learnt the mean gene expression profile, as the R squared is almost above 0.98 for all test conditions *with respect to the data undergoing encoding and decoding*\n",
    "- However, with respect to the ground truth data (what is reported throughout the manuscript), the R squared is on average aroud 0.9. Thus, the encoder-decoder architecture has to be improved. \n",
    "- The energy distance is less sensitive to the effect of the encoder-decoder, as it filters out the noise. Hence, the distances between CellFlow's predictions and the ground truth are more similar to the distances between CellFlow's predictions and the ground truth going through the encoder-decoder.\n",
    "\n",
    "All in all, we should use a more powerful encoder-decoder here. One way to do this is increasing the number of principal components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b0421fa0-cda6-448f-ac37-538a0d977710",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>energy_distance</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Crizotinib</th>\n",
       "      <td>6.929597</td>\n",
       "      <td>0.992093</td>\n",
       "      <td>Panobinostat+Crizotinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Curcumin</th>\n",
       "      <td>1.341948</td>\n",
       "      <td>0.998179</td>\n",
       "      <td>Panobinostat+Curcumin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+SRT1720</th>\n",
       "      <td>2.101867</td>\n",
       "      <td>0.997297</td>\n",
       "      <td>Panobinostat+SRT1720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Sorafenib</th>\n",
       "      <td>1.533446</td>\n",
       "      <td>0.998002</td>\n",
       "      <td>Panobinostat+Sorafenib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRT2104+Alvespimycin</th>\n",
       "      <td>12.955380</td>\n",
       "      <td>0.983267</td>\n",
       "      <td>SRT2104+Alvespimycin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control+Alvespimycin</th>\n",
       "      <td>11.986608</td>\n",
       "      <td>0.984421</td>\n",
       "      <td>control+Alvespimycin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control+Dacinostat</th>\n",
       "      <td>5.757614</td>\n",
       "      <td>0.992954</td>\n",
       "      <td>control+Dacinostat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         energy_distance  r_squared                condition\n",
       "Panobinostat+Crizotinib         6.929597   0.992093  Panobinostat+Crizotinib\n",
       "Panobinostat+Curcumin           1.341948   0.998179    Panobinostat+Curcumin\n",
       "Panobinostat+SRT1720            2.101867   0.997297     Panobinostat+SRT1720\n",
       "Panobinostat+Sorafenib          1.533446   0.998002   Panobinostat+Sorafenib\n",
       "SRT2104+Alvespimycin           12.955380   0.983267     SRT2104+Alvespimycin\n",
       "control+Alvespimycin           11.986608   0.984421     control+Alvespimycin\n",
       "control+Dacinostat              5.757614   0.992954       control+Dacinostat"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metrics_recon_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b981539e-2738-46f1-b820-c8fb740e38a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>energy_distance</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Crizotinib</th>\n",
       "      <td>7.113812</td>\n",
       "      <td>0.871772</td>\n",
       "      <td>Panobinostat+Crizotinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Curcumin</th>\n",
       "      <td>1.363731</td>\n",
       "      <td>0.900114</td>\n",
       "      <td>Panobinostat+Curcumin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+SRT1720</th>\n",
       "      <td>2.104576</td>\n",
       "      <td>0.925166</td>\n",
       "      <td>Panobinostat+SRT1720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Panobinostat+Sorafenib</th>\n",
       "      <td>1.554319</td>\n",
       "      <td>0.893161</td>\n",
       "      <td>Panobinostat+Sorafenib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRT2104+Alvespimycin</th>\n",
       "      <td>14.041352</td>\n",
       "      <td>0.899901</td>\n",
       "      <td>SRT2104+Alvespimycin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control+Alvespimycin</th>\n",
       "      <td>13.771369</td>\n",
       "      <td>0.884888</td>\n",
       "      <td>control+Alvespimycin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control+Dacinostat</th>\n",
       "      <td>5.836791</td>\n",
       "      <td>0.957439</td>\n",
       "      <td>control+Dacinostat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         energy_distance  r_squared                condition\n",
       "Panobinostat+Crizotinib         7.113812   0.871772  Panobinostat+Crizotinib\n",
       "Panobinostat+Curcumin           1.363731   0.900114    Panobinostat+Curcumin\n",
       "Panobinostat+SRT1720            2.104576   0.925166     Panobinostat+SRT1720\n",
       "Panobinostat+Sorafenib          1.554319   0.893161   Panobinostat+Sorafenib\n",
       "SRT2104+Alvespimycin           14.041352   0.899901     SRT2104+Alvespimycin\n",
       "control+Alvespimycin           13.771369   0.884888     control+Alvespimycin\n",
       "control+Dacinostat              5.836791   0.957439       control+Dacinostat"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metrics_true_pred"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellflow",
   "language": "python",
   "name": "cellflow"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
